### ISSUE TITLE
Unable to resolve service for type 'FluentDispatch.Resolvers.FuncResolver'

### ISSUE THREAD

**Issue by lol768** [bbougot/FluentDispatch#4]
I'm getting:

```
System.InvalidOperationException: Unable to resolve service for type 'FluentDispatch.Resolvers.FuncResolver`1[System.Net.IPAddress]' while attempting to activate 'FluentDispatch.Clusters.Cluster`1[System.Net.IPAddress]'.
```

<details>
<summary>Full stack trace</summary>
<pre>
Unhandled exception. System.AggregateException: Some services are not able to be constructed (Error while validating the service descriptor 'ServiceType: FluentDispatch.Clusters.ICluster`1[System.Net.IPAddress] Lifetime: Singleton ImplementationType: FluentDispatch.Clusters.Cluster`1[System.Net.IPAddress]': Unable to resolve service for type 'FluentDispatch.Resolvers.FuncResolver`1[System.Net.IPAddress]' while attempting to activate 'FluentDispatch.Clusters.Cluster`1[System.Net.IPAddress]'.)
 ---> System.InvalidOperationException: Error while validating the service descriptor 'ServiceType: FluentDispatch.Clusters.ICluster`1[System.Net.IPAddress] Lifetime: Singleton ImplementationType: FluentDispatch.Clusters.Cluster`1[System.Net.IPAddress]': Unable to resolve service for type 'FluentDispatch.Resolvers.FuncResolver`1[System.Net.IPAddress]' while attempting to activate 'FluentDispatch.Clusters.Cluster`1[System.Net.IPAddress]'.
 ---> System.InvalidOperationException: Unable to resolve service for type 'FluentDispatch.Resolvers.FuncResolver`1[System.Net.IPAddress]' while attempting to activate 'FluentDispatch.Clusters.Cluster`1[System.Net.IPAddress]'.
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.CreateArgumentCallSites(Type serviceType, Type implementationType, CallSiteChain callSiteChain, ParameterInfo[] parameters, Boolean throwIfCallSiteNotFound)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.CreateConstructorCallSite(ResultCache lifetime, Type serviceType, Type implementationType, CallSiteChain callSiteChain)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.TryCreateExact(ServiceDescriptor descriptor, Type serviceType, CallSiteChain callSiteChain, Int32 slot)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.GetCallSite(ServiceDescriptor serviceDescriptor, CallSiteChain callSiteChain)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.ValidateService(ServiceDescriptor descriptor)
   --- End of inner exception stack trace ---
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.ValidateService(ServiceDescriptor descriptor)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider..ctor(IEnumerable`1 serviceDescriptors, ServiceProviderOptions options)
   --- End of inner exception stack trace ---
   at Microsoft.Extensions.DependencyInjection.ServiceProvider..ctor(IEnumerable`1 serviceDescriptors, ServiceProviderOptions options)
   at Microsoft.Extensions.DependencyInjection.ServiceCollectionContainerBuilderExtensions.BuildServiceProvider(IServiceCollection services, ServiceProviderOptions options)
   at Microsoft.Extensions.DependencyInjection.DefaultServiceProviderFactory.CreateServiceProvider(IServiceCollection containerBuilder)
   at Microsoft.Extensions.Hosting.Internal.ServiceFactoryAdapter`1.CreateServiceProvider(Object containerBuilder)
   at Microsoft.Extensions.Hosting.HostBuilder.CreateServiceProvider()
   at Microsoft.Extensions.Hosting.HostBuilder.Build()
</pre>
</details>

In `Startup.cs#ConfigureServices`:

```c#
services.AddHttpClient();
services.ConfigureCluster(clusterOptions =>
    {
        // We setup 5 local nodes
        clusterOptions.ClusterSize = 5;
    },
    circuitBreakerOptions => { });
services.AddCluster<IPAddress>(
    sp => new IpResolver(sp.GetService<IHttpClientFactory>(), sp.GetService<ILoggerFactory>()));
```

`IpResolver` is taken almost verbatim from the README:

```c#
internal sealed class IpResolver : Resolver<IPAddress>
{
    private readonly ILogger _logger;
    private readonly HttpClient _httpClient;

    public IpResolver(IHttpClientFactory httpClientFactory, ILoggerFactory loggerFactory)
    {
        _httpClient = httpClientFactory.CreateClient();
        _logger = loggerFactory.CreateLogger<IpResolver>();
    }

    /// <summary>
    /// Process each new IP address and resolve its geolocation
    /// </summary>
    /// <param name="item"><see cref="KeyValuePair"/></param>
    /// <param name="nodeMetrics"><see cref="NodeMetrics"/></param>
    /// <param name="cancellationToken"><see cref="CancellationToken"/></param>
    /// <returns><see cref="Task"/></returns>
    protected override async Task Process(IPAddress item,
        NodeMetrics nodeMetrics,
        CancellationToken cancellationToken)
    {
        _logger.LogInformation(
            $"New IP {item.ToString()} received, trying to resolve geolocation from node {nodeMetrics.Id}...");
        var response = await _httpClient.GetAsync(new Uri($"http://ip-api.com/json/{item.ToString()}"),
            cancellationToken);
        _logger.LogInformation("received response " + response.StatusCode);
    }
}
```

v1.0.0 on NuGet., netcoreapp3.0.

**Comment by lol768** [comment]
I seem to be able to bypass this with:

```
            services.AddSingleton<FuncResolver<IPAddress>>(sp => new IpResolver( sp.GetService<IHttpClientFactory>(), sp.GetService<ILoggerFactory>()));
```

Presumably the `AddCluster` extension method should do this though?

**Comment by interviews-dotnet** [comment]
Indeed there was a dependency injection issue, you were right on the suggestion.
I've just fixed it in 35d5cda

**Comment by bbougot** [comment]
Indeed there was a dependency injection issue, you were right on the suggestion.
I've just fixed it in 35d5cda.
Thanks for the report.

**Comment by lol768** [comment]
Thanks a bunch!

**Comment by lol768** [comment]
Can you lmk when this goes out to NuGet? Can clean up my Startup code then :)

**Comment by bbougot** [comment]
No problem I'm going to publish 1.0.1 in the next few hours ;)

**Comment by bbougot** [comment]
@lol768 
I've just published version v1.1.0
It now supports .NET Core 3.1 and there's a slight change in the FluentDispatchClusterHost and FluentDispatchNodeHost which now let configure the logging and monitoring sub-system.
You can see an example [here](https://github.com/bbougot/FluentDispatch/blob/master/Samples/Remote/FluentDispatch.Node/Program.cs#L15) and [here](https://github.com/bbougot/FluentDispatch/blob/master/Samples/Remote/FluentDispatch.Cluster/Program.cs#L18). Samples have been updated.
